src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1401 @@
+  };
+
+  kj::MutexGuarded<Cache> cache;
+};
+
+// The static module bundle maintains an internal table of specifiers to resolve callbacks
+// in memory. Instances must be thread-safe.
+class StaticModuleBundle final : public ModuleBundle {
+public:
+  StaticModuleBundle(
+      Type type,
+      kj::HashMap<Url, ModuleBundle::Builder::ResolveCallback> modules,
+      kj::HashMap<Url, Url> aliases)
+      : ModuleBundle(type),
+        modules(kj::mv(modules)),
+        aliases(kj::mv(aliases)) {}
+
+  kj::Maybe<Module&> resolve(const ResolveContext& context) override {
+    KJ_IF_SOME(aliased, aliases.find(context.specifier)) {
+      // The specifier is registered as an alias. We need to resolve the alias instead
+      ResolveContext newContext {
+        .type = context.type,
+        .source = context.source,
+        .specifier = aliased,
+        .referrer = context.referrer,
+        .rawSpecifier = context.rawSpecifier,
+      };
+      return resolve(newContext);
+    }
+
+    auto lock = cache.lockExclusive();
+    KJ_IF_SOME(cached, lock->find(context.specifier)) {
+      return checkModule(context, *cached);
+    }
+
+    // Module was not cached, try to resolve it.
+    KJ_IF_SOME(found, modules.find(context.specifier)) {
+      KJ_IF_SOME(resolved, found(context)) {
+        Module& module = *resolved;
+        lock->upsert(context.specifier.clone(), kj::mv(resolved));
+        return checkModule(context, module);
+      }
+    }
+
+    return kj::none;
+  }
+
+private:
+  kj::HashMap<Url, ModuleBundle::Builder::ResolveCallback> modules;
+  kj::HashMap<Url, Url> aliases;
+  kj::MutexGuarded<kj::HashMap<Url, kj::Own<Module>>> cache;
+};
+
+kj::HashSet<kj::StringPtr> toHashSet(kj::ArrayPtr<const kj::String> arr) {
+  kj::HashSet<kj::StringPtr> set;
+  set.insertAll(arr);
+  return kj::mv(set);
+}
+
+}  // namespace
+
+// ======================================================================================
+kj::Own<ModuleBundle> ModuleBundle::newFallbackBundle(Builder::ResolveCallback callback) {
+  return kj::heap<FallbackModuleBundle>(kj::mv(callback));
+}
+
+void ModuleBundle::getBuiltInBundleFromCapnp(
+      BuiltinBuilder& builder,
+      Bundle::Reader bundle,
+      BuiltInBundleOptions options) {
+
+  static constexpr auto checkAllowDataModules = [](BuiltInBundleOptions options) -> bool {
+    if ((options & BuiltInBundleOptions::ALLOW_DATA_MODULES) !=
+            BuiltInBundleOptions::ALLOW_DATA_MODULES) {
+      LOG_ERROR_ONCE("Builtin wasm module or data module used without passing "
+                     "builtin-wasm-modules autogate flag");
+      return false;
+    }
+    return true;
+  };
+
+  auto filter = ([&] {
+    switch (builder.type()) {
+      case Module::Type::BUILTIN: return ModuleType::BUILTIN;
+      case Module::Type::BUILTIN_ONLY: return ModuleType::INTERNAL;
+      case Module::Type::BUNDLE: break;
+      case Module::Type::FALLBACK: break;
+    }
+    KJ_UNREACHABLE;
```

Comment:
So... we get to unreachable for non-builtin modules? And we check that `module.getType()` is equal to the lambda? Can't we just check if it is equal to BUILTIN or BUILTIN_ONLY?

Reply:
we could. I kind of prefer this one tho. Works either way I suppose. For now I'll keep it as is.